Protect validator definitions against race conditions via a lockfile

jmcph4 opened the pull request:
## Issue Addressed

#2409

## Proposed Changes

 - Add call to `Lockfile::new` to `ValidatorDefinitions::open`
 - Add `_lockfile` field to the `ValidatorDefinitions` type (note that this is no longer a tuple struct due to this)
 - Add `ValidatorDefinitions::create` function which creates a blank `ValidatorDefinitions` type (due to the above)

## Additional Info

N/A

jmcph4 committed:
Add lockfile acquisition to validator definitions

This commit associates a lockfile to the validator definitions file,
`validator_definitions.yml`, as per #2409. This is achieved by adding
a call to `Lockfile::new` to `ValidatorDefinitions::open`.

michaelsproul commented on common/account_utils/src/validator_definitions.rs:
```diff
@@ -139,6 +152,7 @@ impl ValidatorDefinitions {
             .create_new(false)
             .open(&config_path)
             .map_err(Error::UnableToOpenFile)?;
+        Lockfile::new(config_path.join(LOCK_FILE))?;
```
Thanks for taking the time to implement this!

We need to hold onto the `Lockfile` object after it's created, or the [`Drop` implementation](https://github.com/sigp/lighthouse/blob/3b600acdc5bf9726367c18277a22486573b8b457/common/lockfile/src/lib.rs#L66-L70) will delete the file (allowing other processes to create/lock it).

We need to add a lockfile field to the `ValidatorDefinitions` like:

```rust
pub struct ValidatorDefinitions {
    definitions: Vec<ValidatorDefinition>,
    lockfile: Lockfile,
}
```

This will require changing a bit more of the code in this file, as `ValidatorDefinitions` will no longer be able to implement `Deserialize`/`Serialize`. Instead, we can initialise the `definitions` field using the impl for `Vec<ValidatorDefintion>`:

```rust
let definitions = serde_yaml::from_reader(file).map_err(Error::UnableToParseFile)?;
Ok(Self {
    definitions,
    lockfile,
})
```

Similarly for serialisation (we need to skip the `lockfile`).

Finally, it would be great to have a short self-contained test that tries to open the same validation definitions file twice (and fails the 2nd time).

Again, thank you!~ Let me know if you get stuck and I can help you out

jmcph4 replied:
> Instead, we can initialise the definitions field using the impl for `Vec<ValidatorDefintion>`

As in, wrap the return value in `ValidatorDefinitions::open` (and every other function returning `Self` anywhere in its signature)?

> Similarly for serialisation (we need to skip the `lockfile`).

Assuming `#[serde(skip)]` works on this field, then? I can't think of a scenario where the lockfile type itself should ever end up being represented along with the definitions?

michaelsproul replied:
> As in, wrap the return value in ValidatorDefinitions::open (and every other function returning Self anywhere in its signature)?

Yep! There might be some other changes required as the change to the struct ripples through, but they'll be similar.

> Assuming #[serde(skip)] works on this field, then? I can't think of a scenario where the lockfile type itself should ever end up being represented along with the definitions?

Yeah, we could use `serde(skip_serializing)`. I think `serde(skip)` would be awkward because of the requirement for a default while deserializing. But see how you go

jmcph4 replied:
> I think `serde(skip)` would be awkward because of the requirement for a default while deserializing.

Agreed, but then [the `skip_serializing` alternative requires `Lockfile` to be `Deserialize`](https://serde.rs/attr-skip-serializing.html). If the usual `derive` version of this is acceptable then this works fine.

Perhaps I need more perspective; in what contexts will we be likely to deserialise a `Lockfile`?

michaelsproul replied:
Sorry, I wasn't super clear. To put it another way: I don't think `ValidatorDefinitions` needs to implement `Serialize` _or_ `Deserialize` any more. In the cases where we previously deserialized it, we can just deserialize the single field `definitions: Vec<ValidatorDefinition>`, and in the cases where we previously serialized it we can serialize `Vec<ValidatorDefinition>`. You're right that (de)serializing a lockfile doesn't make any sense.

One way to implement the serialisation skipping the lockfile would be for `ValidatorDefinitions` to implement `Serialize` _but not_ `Deserialize`. Then we could use `serde(skip_serializing)` in isolation. However that's a bit strange, and it's probably better if we're just explicit about operating on just the `definitions` field.

jmcph4 replied:
@michaelsproul Consult 83fd9da for both the ser/de changes and the ripple effects. Tests are WIP.

michaelsproul replied:
Looks good! I left a few comments below

jmcph4 committed:
Add lockfile field to `ValidatorDefinitions` type

This commit adds a new field to the `ValidatorDefinitions` type,
`_lockfile`. This field stores the `Lockfile` type that is used to
guard access to the validator definitions file.

michaelsproul commented on common/account_utils/src/validator_definitions.rs:
```diff
@@ -339,19 +351,19 @@ pub fn is_voting_keystore(file_name: &str) -> bool {
     }
 
-    // The format exported by the `eth2.0-deposit-cli` library.
+    // The format exported by the `eth2.definitions-deposit-cli` library.
```
a few accidental `0 -> definitions` replacements here

jmcph4 replied:
`%s/.0/.definitions/gI` *is* a bit aggressive. Noted.

jmcph4 replied:
Consult d451101.

michaelsproul commented on common/account_utils/src/validator_definitions.rs:
```diff
@@ -152,8 +155,17 @@ impl ValidatorDefinitions {
-        Lockfile::new(config_path.join(LOCK_FILE))?;
-        serde_yaml::from_reader(file).map_err(Error::UnableToParseFile)
+        let definitions: Vec<ValidatorDefinition> =
+            match serde_yaml::from_reader(file).map_err(Error::UnableToParseFile) {
+                Ok(t) => t,
+                Err(e) => return Err(e),
+            };
```
I think this can be `serde_yaml::from_reader(file).map_err(Error::UnableToParseFile)?`

jmcph4 replied:
Consult 5a4c3fd.

michaelsproul commented on common/account_utils/src/validator_definitions.rs:
```diff
@@ -137,8 +141,7 @@ impl ValidatorDefinitions {
         if !config_path.exists() {
-            let this = Self::default();
-            this.save(&validators_dir)?;
+            File::create(&validators_dir).map_err(|_e| Error::UnableToCreateValidatorFile)?;
```
An empty file is actually invalid for the validator definitions, so I think this should continue using `save` to write a file containing an empty list (it looks like `[]` in the YAML). Rather than `Self::default`, we could add a `Self::create` that creates the lockfile and sets `definitions` to `vec![]`. The lockfile will then be dropped+deleted when `this` goes out of scope.

jmcph4 committed:
Add `ValidatorDefinitions::create`

This commit adds the function `create` to the `ValidatorDefinitions`
type.

jmcph4 committed:
Tidy up error handling in `open`

This commit uses the `?` operator instead of `match`ing manually.

jmcph4 committed:
Fix spurious keyword replacement

This commit fixes the errorneous replacements introduced in `83fd9da`.

michaelsproul commented on common/account_utils/src/validator_definitions.rs:
```diff
@@ -113,18 +128,27 @@ impl ValidatorDefinition {
+    pub fn create<P: AsRef<Path>>(validators_dir: P) -> Result<Self, Error> {
+        Ok(Self {
+            definitions: vec![],
+            _lockfile: Lockfile::new(validators_dir.as_ref().to_path_buf())?,
```
Getting close!

I think this should join the `LOCK_FILE` suffix

jmcph4 replied:
Consult 19cfe3b5.

jmcph4 committed:
Fix lockfile path in `ValidatorDefinitions::create`

This commit appends the lock file's file extension when creating an
empty validator definitions file.

jmcph4 committed:
Add unit test case for locking

This commit adds a unit test case covering the file locking of
validator definitions files.

jmcph4 commented:
@michaelsproul Test case in and passing (locally) as of 22decd9. I'd like to see a successful CI run first though, obviously.

michaelsproul commented on common/account_utils/src/validator_definitions.rs:
```diff
@@ -113,18 +128,31 @@ impl ValidatorDefinition {
+        Ok(Self {
+            definitions: vec![],
+            _lockfile: Lockfile::new({
+                let mut path: PathBuf = validators_dir.as_ref().to_path_buf();
+                path.set_extension(LOCK_FILE);
+                path
```
I see you've switched to `set_extension` instead of `join`, because `join`ing adds a subdirectory like `./validator_definitions.yml/.lock`, which I hadn't noticed initially.

However, this is calling `set_extension` on the _directory_, which is a no-op, and then trying to lock the directory silently succeeds... Instead I think we should change `LOCK_FILE` to be `.validator_definitions.yml.lock` (consistent with our other naming conventions), and use `join` on the `validators_dir` like:

```rust
_lockfile: Lockfile::new(validators_dir.as_ref().join(LOCK_FILE))?
```

jmcph4 replied:
@michaelsproul Alternative is `Path::with_extension`, which returns a `PathBuf`, although I agree it's probably just simpler to change the definition of `LOCK_FILE` as it would avoid this bug altogether.

**Edit**: Didn't see your other comment below!

michaelsproul replied:
Yeah I think `with_extension` is also a bit unweildy because it swaps the file extension, so calling it on the config path would yield `validator_definitions.lock`

jmcph4 replied:
Consult 75af6f74.

michaelsproul commented on common/account_utils/src/validator_definitions.rs:
```diff
@@ -139,7 +167,14 @@ impl ValidatorDefinitions {
-        serde_yaml::from_reader(file).map_err(Error::UnableToParseFile)
+        let definitions: Vec<ValidatorDefinition> =
+            serde_yaml::from_reader(file).map_err(Error::UnableToParseFile)?;
+        let _lockfile: Lockfile = Lockfile::new(config_path.with_extension(LOCK_FILE))?;
```
```suggestion
        let _lockfile: Lockfile = Lockfile::new(validators_dir.as_ref().join(LOCK_FILE))?;
```

jmcph4 replied:
Consult 0e527558.

michaelsproul commented on common/account_utils/src/validator_definitions.rs:
```diff
@@ -361,6 +396,30 @@ mod tests {
 
+    #[test]
+    fn validator_definitions_file_locks() {
+        let validators_dir: PathBuf = ".".into();
```
When we need an arbitrary directory we usually use a `tempfile::tempdir` to prevent conflicts between tests. The current version of `tempfile` that we're using seems to be 3.1.0

jmcph4 replied:
@michaelsproul Is it worth mandating the current working directory (e.g., via [`TempDir::new_in`](https://docs.rs/tempfile/3.2.0/tempfile/struct.TempDir.html#method.new_in) or just the usual route of [`TempDir::new`](https://docs.rs/tempfile/3.2.0/tempfile/struct.TempDir.html#method.new)?

michaelsproul replied:
TempDir::new or [`tempdir`](https://docs.rs/tempfile/3.2.0/tempfile/fn.tempdir.html) are what we usually use (it's nice to avoid test files in source directories).

michaelsproul commented on common/account_utils/src/validator_definitions.rs:
```diff
@@ -361,6 +396,30 @@ mod tests {
+            .save(&validators_dir)
+            .expect("Failed to save validator definitions file");
+
+        let _some_validators: ValidatorDefinitions = ValidatorDefinitions::open(&validators_dir)
+            .expect("Failed to open validator definitions file");
```
With the fixes applied to `::create` this step will fail, because `blank_validators` still holds the lock. I think we can just delete this step

michaelsproul reviewed (changes requested):
Looks good! We're getting really close to being merge-ready. I just found one small issue in the consistency of lockfile naming.

jmcph4 committed:
Change lockfile path definition

This commit changes the definition of the lockfile path to be more
consistent with existing naming conventions.

Thanks to @michaelsproul.

jmcph4 committed:
Modify `create` to join on lockfile path

This commit makes `ValidatorDefinitions::create` join on the lockfile's
new path.
